### PR TITLE
Exercise 5

### PR DESCRIPTION
**Pull request by shaylipatel**
Module 5
A single, multi-stage Dockerfile is used to specify development and production containers. 
A production container uses a production-ready server (gunicorn) to run the app.
The README contains instructions on how to build and run development and production containers. 
No secrets are included in the built images.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+FROM python:3.12-rc-alpine as base
+RUN apk update
+RUN apk add curl
```
Comment (JackMeadDev) [inline Dockerfile]: Minor but you will sometimes see advice to combine layers where appropriate as it creates a slightly more optimised image, so e.g. here we might use:
```suggestion
RUN apk update && apk add curl
```

File: README.md [hunk 2]
```diff
@@ -125,3 +125,33 @@ On the Ansible Control Node, run the following command to run it in the interact
...
+$ docker build --target development --tag todo-app:dev .
+```
+or the following for prod:
+```bash
+$ docker build --target production --tag todo-app:prod .
+```
+
+You can then start the dev container by running:
+```bash
+$ docker run --env-file ./.env -p 5100:80 --mount "type=bind,source=$(pwd)/todo_app,target=/app/todo_app" todo-app:dev
```
Comment (JackMeadDev) [inline README.md]: This doesn't quite work out the box for me; I think you've got a couple of arguments mixed with the production ones:
```suggestion
$ docker run --env-file ./.env -p 5100:5000 --mount "type=bind,source=$(pwd)/todo_app,target=/todo_app/todo_app" todo-app:dev
```
(Note that you've chosen (sensibly!) a `todo_app` folder for your working directory, and we then copy another directory in also called `todo_app`, so we end up with the slightly funny looking target of `/todo_app/todo_app`)

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+RUN curl -sSL https://install.python-poetry.org | python3 -
+ENV PATH="${PATH}:/root/.local/bin"
+WORKDIR /todo_app
+COPY pyproject.toml poetry.lock ./
+RUN poetry install 
+EXPOSE 5000
+COPY todo_app ./todo_app
+FROM base as production
+ENTRYPOINT ["poetry" , "run"]
+CMD ["gunicorn",  "--bind", "0.0.0.0:5000", "todo_app.app:create_app()"]
```
Comment (JackMeadDev) [inline Dockerfile:13]: This fails with `Command not found: gunicorn` for me - have you run `poetry add gunicorn` locally? If so, have you forgotten to check in the changes to `pyproject.toml` and `poetry.lock` possibly?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+RUN apk update
+RUN apk add curl
+RUN curl -sSL https://install.python-poetry.org | python3 -
+ENV PATH="${PATH}:/root/.local/bin"
+WORKDIR /todo_app
+COPY pyproject.toml poetry.lock ./
+RUN poetry install 
+EXPOSE 5000
+COPY todo_app ./todo_app
+FROM base as production
```
Comment (JackMeadDev) [inline Dockerfile]: Minor, but a couple of small line breaks just to separate out the different sections would probably make this easier to read, e.g. separating the stages:
```suggestion
COPY todo_app ./todo_app

FROM base as production
```

**Review by JackMeadDev — APPROVED** [review]
Nice work, I was able to run your app in both dev & production containers without difficulty, although neither worked perfectly out of the box - so worth making sure you've got accurate commands and that `gunicorn` is listed in your project dependencies before merging! 😄

File: Dockerfile (hunk 1 above)
Comment (shaylipatel) [reply Dockerfile]: Changes made

File: Dockerfile (hunk 3 above)
Comment (shaylipatel) [reply Dockerfile:13]: have committed the pyproject.toml and poetry.lock file

File: README.md (hunk 2 above)
Comment (shaylipatel) [reply README.md]: Updated the README.md instructions

**Pull request merged**